### PR TITLE
fix: navbar layout issues

### PR DESCRIPTION
**Pull request by joaoviana**
Closes: #7531 #8364

### Description:

When users were editing charts from dashboards, they would see an issue with the Page layout where the Explorer sidebar and the whole `Page` seemed cropped. 
The reason why this is happening currently if because: 
- the main nav bar behaves one way, has specific height, styling, etc
- the blue banner that we render when `editing charts from dashboards` is a different component, with less of the styling we apply to the main nav bar

Instead of replacing the `Navbar` completely when that's the case, we can just render different content inside the same component.

What was done: 
- Created separate components for: `main nav bar content`, `dashboard explorer banner`, and the `preview banner`
- On the `index.tsx` of `NavBar`, we have now a Nav bar chooser hook, that controls the content & styles of the Header. 
- Consequently, we now don't need the `hasBanner` prop for the `Page` component, but we **still** let the Page component apply styles when the project is PREVIEW.

Here is a screen recording:
https://github.com/lightdash/lightdash/assets/7611706/8c041466-ea8a-4e3a-8ca7-2cec775f9812

### Reviewer actions

- [ ] I have manually tested the changes in the preview environment
- [ ] I have reviewed the code
- [ ] I understand that "request changes" will block this PR from merging

### CODE REVIEW

**Comment by joaoviana** [comment]
https://github.com/lightdash/lightdash/assets/7611706/8ae01e82-6b21-490a-8feb-b41ceaa28203

**Comment by IrakliJani** [comment]
I am not sure what is wrong with the render environment. I re-launched the headless browser, but the tests won't start. You can close the PR and reopen it.

**Comment by IrakliJani** [comment]
looks great in all of the edge cases that I know of...

even this one 😱 

CleanShot 2024-03-19 at 00 30 51@2x

**Comment by joaoviana** [comment]
E2E tests passed @IrakliJani , thanks for mentioning this issue :)

File: packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    const { getEditingDashboardInfo, clearDashboardStorage } =
+        useDashboardStorage();
+    const { name: dashboardName, dashboardUuid } = getEditingDashboardInfo();
+
+    useEffect(() => {
+        window.addEventListener('unload', clearDashboardStorage);
+        return () =>
+            window.removeEventListener('unload', clearDashboardStorage);
+    }, [clearDashboardStorage]);
```
Comment (IrakliJani) [inline packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx:27]: ok im little bit lost here

File: packages/frontend/src/components/NavBar/index.tsx [hunk 2]
```diff
@@ -1,213 +1,111 @@
...
+    );
     const headerContainerHeight =
         NAVBAR_HEIGHT + (isCurrentProjectPreview ? BANNER_HEIGHT : 0);
 
+    const renderNavBarContent = () => {
+        switch (navBarMode) {
+            case NavBarMode.EDITING_DASHBOARD_CHART:
+                return <DashboardExplorerBanner projectUuid={projectUuid} />;
+            case NavBarMode.DEFAULT:
+            default:
```
Comment (IrakliJani) [inline packages/frontend/src/components/NavBar/index.tsx:84]: navbarMode can never be undefined so you can move the content from `default` block to NavbarMode.DEFAULT and call `assertUnreachable` inside `default` block

**Review by IrakliJani — APPROVED** [review]
this is great. one question. otherwise 👍

File: packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx (hunk 1 above)
Comment (joaoviana) [reply packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx:27]: Same 😆 
This was here before, so it clears the dashboardStorage when the banner unloads. This usage is not recommended, but will address on a separate PR

File: packages/frontend/src/components/NavBar/index.tsx (hunk 2 above)
Comment (joaoviana) [reply packages/frontend/src/components/NavBar/index.tsx:84]: Yes 🤦 
Thanks! Will add this 👍

File: packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx (hunk 1 above)
Comment (IrakliJani) [reply packages/frontend/src/components/NavBar/DashboardExplorerBanner.tsx:27]: oh cool 😆

File: packages/frontend/src/components/NavBar/index.tsx (hunk 2 above)
Comment (IrakliJani) [reply packages/frontend/src/components/NavBar/index.tsx]: perfect 👍
